src/planet-a/SubdivTriangleMesh.h

```diff
@@ -0,0 +1,501 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include "SubdivSkeleton.h"
+
+#include <osp/types.h>
+
+#include <memory>
+#include <variant>
+
+namespace planeta
+{
+
+// ID for all chunks, from 0 to m_chunkMax
+enum class ChunkId : uint16_t {};
+
+// ID for all shared vertices; from 0 to m_sharedMax
+enum class SharedVrtxId : uint32_t {};
+
+// Indices of vertices, unaware of vertex size;
+// from 0 to (m_chunkMax*m_chunkVrtxCount + m_sharedMax)
+enum class VertexId : uint32_t {};
+
+// IDs for any chunk's shared vertices; from 0 to m_chunkVrtxSharedCount
+enum class ChunkLocalSharedId : uint16_t {};
+
+// IDs for any chunk's fill vertices; from 0 to m_chunkVrtxSharedCount
+enum class ChunkLocalFillId : uint16_t {};
+
+using SharedVrtxStorage_t = osp::IdRefCount<SharedVrtxId>::Storage_t;
+
+//-----------------------------------------------------------------------------
+
+/**
+ * Convert XY coordinates to a triangular number index
+ *
+ *  0
+ *  1  2
+ *  3  4  5
+ *  6  7  8  9
+ * 10 11 12 13 14
+ * ...
+ * x = right; y = down; (0, 0) = 0
+ *
+ * @param x [in]
+ * @param y [in]
+ * @return triangular number
+ */
+constexpr int xy_to_triangular(int x, int y) noexcept
+{
+    return ( y * (y + 1) ) / 2 + x;
+};
+
+constexpr std::pair<ChunkLocalSharedId, bool> coord_to_shared(
+        uint16_t x, uint16_t y, uint16_t chunkEdgeVrtxCount) noexcept
+{
+    // Tests if (x,y) lies along right, bottom, or left edges of triangle
+    // branchless because why not
+    bool const lft = (x == 0);
+    bool const btm = !lft && (y == chunkEdgeVrtxCount);
+    bool const rte = !btm && (x == y);
+
+    // Calculate chunk local shared ID. starts at 0 from the top of the chunk,
+    // and increases counterclockwise along the edge
+    uint16_t const offset = lft * y
+                          + btm * (chunkEdgeVrtxCount + x)
+                          + rte * (chunkEdgeVrtxCount * 3 - x);
+
+    return {ChunkLocalSharedId(offset), rte || btm || lft};
+}
+
+//-----------------------------------------------------------------------------
+
+
+struct Chunk
+{
+    SkTriStorage_t m_skeletonTri;
+};
+
+class ChunkedTriangleMeshInfo
```

Comment:
Any reason why this is all defined within the class declaration? I find it easier to understand the class at a glance if it's not packed with implementation details. You can still keep it as header-only, but just put the implementations down at the bottom with an explicit `inline` specifier. Just personal preference.

Reply:
Too lazy. It's all unused at the moment.